Ekr editorial 17 3 (#2164)
quicwg/base-drafts: draft-ietf-quic-transport.md

smaller datagrams.  Sending padded datagrams ensures that the server is not
overly constrained by the amplification restriction.

Packet loss, e.g., of the server's Handshake packet, can cause a

> situation in which the server cannot send becaus of the

Reviewer: ```suggestion
situation in which the server cannot send because of the
```

A token allows a server to correlate activity between the connection where the
token was issued and any connection where it is used.  Clients that want to
break continuity of identity with a server MAY discard tokens provided using the
NEW_TOKEN frame.  Tokens obtained in Retry packets MUST NOT be discarded
during connection establishment (they will not be used with new connections).

> 

Reviewer: How about a positive: A token obtained in a Retry packet must be used immediately during the connection attempt and cannot be used in subsequent connection attempts.

Reviewer 2: What @mikkelfj said.

The parenthetical doesn't really make sense to me - it's true, but I don't see how it is connected to the primary statement.

between a specific local address and a specific peer address, where an address
is the two-tuple of IP address and port.

Path validation tests that packets (PATH_CHALLENGE) can be both sent

> to and received (PATH_RESPONSE) from a peer on the path.  Importantly,

Reviewer 3: These are not packets.

Reviewer 2: I agree, this likely confuses things more than it helps.

it validates that the packets received from the migrating endpoint do
not carry a spoofed source address.

> 

Reviewer: But that isn't exactly true since MITM can spoof the source address of these packets if it can guess the packet content (as you pointed out).

Author: Sure.. I didn't change this text, I just added the parentheticals to make it clearer. This PR is essentially editorial

would an Initial packet, ensuring that connection migration is no more load on a
new path than establishing a new connection.

The endpoint MUST use unpredictable data in every PATH_CHALLENGE frame

> so that it can associate the peer's response with the causative

Reviewer: ```suggestion
so that it can associate the peer's response with the corresponding
```

## Successful Path Validation

> A new address is considered valid when A PATH_RESPONSE frame is received

Reviewer 4: ```suggestion
A new address is considered valid when a PATH_RESPONSE frame is received
```

that meets the following criteria:

> - It contains that was sent in a previous PATH_CHALLENGE. Receipt of an

Reviewer: ```suggestion
- It contains the data that was sent in a previous PATH_CHALLENGE. Receipt of an
```

> - It contains that was sent in a previous PATH_CHALLENGE. Receipt of an

Reviewer 3: I think you a word out here.

  acknowledgment for a packet containing a PATH_CHALLENGE frame is not adequate
  validation, since the acknowledgment can be spoofed by a malicious peer.

> - It is from the same remote address as that to which the

Reviewer 4: ```suggestion
- It is sent from the same remote address as that to which the
```

>   corresponding PATH_CHALLENGE was sent. If a PATH_RESPONSE frame is

Reviewer: It was received from the remote address that the corresponding PATH_CHALLENGE was sent to.

Author: I think this text is fine

  received from a different remote address than the one to which the
  PATH_CHALLENGE was sent, path validation is considered to have failed,

>   even if the data matches that sent in the PATH_CHALLENGE.

Reviewer: ```suggestion
  even if the PATH_RESPONSE data matches the PATH_CHALLENGE data.
```

Author: I think this is fine

- It is received on the same local address from which the
  corresponding PATH_CHALLENGE was sent.

> Note that receipt on a different local address doesn't result in path

Reviewer: ```suggestion
Note that receipt on a different local address does not result in path
```

Author: Prefer as-is

Reviewer 2: @janaiyengar insists on avoiding contractions.

Reviewer 4: +1. I do.

draining.  A key update might prevent the endpoint from moving from the closing
state to draining, but it otherwise has no impact.

While in the closing period, an endpoint could receive packets from a
new source address, indicating a client connection migration
({{migration}}). An endpoint in the closing state MUST strictly limit
the number of packets it sends to this new address until the address
is validated (see {{migrate-validate}}). A server in the closing state
MAY instead choose to discard packets received from a new source

> address.

Reviewer 2: The change to the fill-column is making this harder to review than I'd like.  I guess that's just a consequence of insisting on wrapping at <80...

Reviewer 3: I tried rewrapping it for him, but that makes all the comments outdated, unfortunately.  I think this will need a rewrap just before merge instead, unfortunately.

## Stateless Reset {#stateless-reset}

> A stateless reset is provided as an option of last resort for a server that

Reviewer 3: No, SR is bilateral, even if we expect servers to use them more frequently.

Author: "A client MUST NOT include an original connection ID, a stateless reset token, or
a preferred address.  A server MUST treat receipt of any of these transport
parameters as a connection error of type TRANSPORT_PARAMETER_ERROR."

It does appear that a client can send one in NCID, but that's a real inconsistency.

Reviewer 3: - OCID is only applicable to servers because it's about Retry
- Preferred Address is only applicable to servers and contains an SRT
- SRT is useless if it's ever exchanged in the clear, which the client's TPs are

After the initial CID, the client can provide an SRT and the behavior is bilateral.  There was a discussion about enabling the client to supply an SRT for the initial CID after the handshake, but there wasn't great appetite for the added complexity.  If the server cares about getting an SRT, it just starts using the next CID as soon as it gets one.

Author: Ugh.

Reviewer 2: Yep :)

Reviewer 4: It'd all be a lot simpler if we just went with client and server :-)

number of random bytes following it are set to unpredictable values.  The last
16 bytes of the datagram contain a Stateless Reset Token.

A stateless reset will be interpreted by a recipient as a packet with
a short header.  For the packet to appear as valid, the Random Bits

> field needs to include at least 182 bits of data (or 24 bytes, less

Reviewer 3: I appreciate that you're removing the suggestion that these should be "random" despite the name of the field.  However, it seems worthwhile to keep "unpredictable" at least.

Author: The previous graf dicates the contents. This graf is about how the recipient will interpret them, and the recipient has no good way of knowing if they are random or unpredictable or whatever.

> the two fixed bits). This is intended to allow for a destination

Reviewer: ```suggestion
the two fixed bits). This is intended to allow for a Destination
```

> connection ID of the maximum length permitted, with a minimal packet

Reviewer: ```suggestion
Connection ID of the maximum length permitted, with a minimal packet
```

Author: I'll let MT weigh in on this. This is existing text.

Reviewer 3: Our convention is that the term "connection ID" is not capitalized, but the fields named "Connection ID" are.

Reviewer 2: This is a field, so the capitalization is correct.

### Looping {#reset-looping}

The design of a Stateless Reset is such that without knowing the
stateless reset token it is indistinguishable from a valid packet.
This means that if a server sends a Stateless Reset to another server,

> that might trigger the sending of a Stateless Reset in response, which

Reviewer 3: Can we get rid of one of these "that"s?

Author: feel free to suggest.

Reviewer 3: "This means that if a server sends a Stateless Reset to another server, it might receive another Stateless Reset in response.  As a result...."

Reviewer 2: One fewer: "If a server sends a Stateless Reset to another server, it might receive another Stateless Reset in response."

includes the connection ID that the sender of the packet wishes to use (see
{{negotiating-connection-ids}}).

> 

Reviewer 3: Extra line seems unnecessary here.

Reviewer 2: ```suggestion
```

Handshake packets are their own packet number space, and thus

> the first Handshake packet sent by a server contains a packet number of 0.

Reviewer 3: Is this a requirement?  Last I knew, some implementations were choosing to just keep a counter incrementing across all packet spaces, and that was okay.  The key point is that (Initial,X) and (Handshake,X) are different packets, but IIRC there's no requirement that each space start at zero.

Author: "   The CRYPTO frame can be sent in different packet number spaces.  The
   sequence numbers used by CRYPTO frames to ensure ordered delivery of
   cryptographic handshake data start from zero in each packet number
   space.
"

Reviewer 3: That's the offset of the CRYPTO stream, not the packet numbers.

Author: My mistake. Here:

https://tools.ietf.org/html/draft-ietf-quic-transport-16#section-12.3

" This enforces cryptographic separation between the data sent in the
   different packet sequence number spaces.  Each packet number space
   starts at packet number 0.  Subsequent packets sent in the same
   packet number space MUST increase the packet number by at least one."

Reviewer 2: I think that we're operating on the assumption that you start sending from 0, but there is no strict requirement because it's unenforceable.  Loss.

Retry packet (that is, the Destination Connection ID field from the client's
first Initial packet) in the transport parameter.

If the client received and processed a Retry packet, it MUST validate
that the original_connection_id transport parameter is present and

> correct; otherwise, it validates that the transport parameter is

Reviewer 4: ```suggestion
correct; otherwise, it MUST validate that the transport parameter is
```

absent.  A client MUST treat a failed validation as a connection error

> of type TRANSPORT_PARAMETER_ERROR.

Reviewer 3: Adding the MUST to the validate-is-present fork seems unnecessary, but if you're going to, isn't it also a MUST to validate-is-absent?

Author: yes I agree. What I'm trying to do is remove language which is apparently descriptive but is actually normative. I.e., "the agent does X". That's a requirement

Reviewer 3: Then "otherwise, it MUST validate that...."

Reviewer 2: I'm OK with the existing or as Mike suggests.

Reviewer 4: Hah, yes. Suggested above.

Reason Phrase Length:

: A variable-length integer specifying the length of the reason phrase

>   in bytes.  Because CONNECTION_CLOSE frame cannot be split between

Reviewer: ```suggestion
  in bytes.  Because a CONNECTION_CLOSE frame cannot be split between
```

specification needs to describe the format and assigned semantics of any fields
in the frame.

> Expert(s) SHOULD be biased towards approving registrations unless

Reviewer 2: Not really an interoperability requirement.

```suggestion
Expert(s) are encouraged to be biased towards approving registrations unless
```
